### DISCUSSION TITLE
Is there interest in automated releases to docker-hub?

### DISCUSSION THREAD

**Discussion by toadkicker** [libvips/libvips#2031] [Ideas]
I was wondering if there is interest in publishing an official build to docker-hub, and if so would a pull request to create it be accepted?

**Comment by jcupitt** [comment]
Hello @toadkicker, 

Yes, this has been talked about a few times, but no one has got around to doing it. A patch would be very welcome!

As I remember, the idea was to push from Travis during CI, but perhaps there's a better way.

**Comment by augustocdias** [comment]
I have a in progress image for it if you want to take it as a starting point

```
FROM alpine

RUN apk update

RUN apk add build-base pkgconfig glib-dev gobject-introspection-dev expat-dev tiff-dev libjpeg-turbo-dev libexif-dev giflib-dev librsvg-dev lcms2-dev libpng orc-dev libwebp-dev libheif-dev

RUN apk add libimagequant-dev --repository=http://dl-cdn.alpinelinux.org/alpine/edge/testing

RUN https://github.com/libvips/libvips/releases/download/v8.8.3/vips-8.8.3.tar.gz

RUN mkdir vips

RUN tar xvzf vips-8.8.3.tar.gz -C vips --strip-components 1

WORKDIR /vips

RUN ./configure

RUN make

RUN make install

RUN ldconfig

WORKDIR /

RUN rm -rf /vips
```

**Comment by toadkicker** [comment]
Made a couple of performance improvements and syntax changes:

```
FROM alpine
ENV VIPSVERSION 8.8.3
RUN apk update
RUN apk add build-base pkgconfig glib-dev gobject-introspection-dev expat-dev tiff-dev libjpeg-turbo-dev libexif-dev giflib-dev librsvg-dev lcms2-dev libpng orc-dev libwebp-dev libheif-dev
RUN apk add libimagequant-dev --repository=http://dl-cdn.alpinelinux.org/alpine/edge/testing

RUN \
  # Build libvips
  cd /tmp && \
  wget https://github.com/libvips/libvips/releases/download/v$VIPSVERSION/vips-$VIPSVERSION.tar.gz && \
  tar zxvf vips-$VIPSVERSION.tar.gz && \
  cd /tmp/vips-$VIPSVERSION && \
  ./configure --enable-debug=no --without-python $1 && \
  make && \
  make install

RUN ldconfig /

CMD [ "/bin/bash" ]
```

**Comment by augustocdias** [comment]
what is the `$1`? How is it passed?

**Comment by toadkicker** [comment]
@augustocdias that is a shell variable; docker supports sending shell arguments to dockerfiles. So we can write a configure script that has sane defaults and args for others.

@jcupitt -Do we want to back support Docker, and if so to what version?
To support that I think we would make a Docker template file, call the Github releases API, iterate through each release and create a Dockerfile for each tag/asset. That means for each release we have a Windows based and Linux based docker container. Then we can roll those out to Dockerhub. I have some draft code here to do this:

```
#!/usr/bin/env bash

# Fetch RELEASEs from the repo
TMPL_PATH='../templates/Dockerfile.template'
LIBVIPS_RELEASES=$(curl -s https://api.github.com/repos/libvips/libvips/releases  | jq '.')
# for each RELEASE, create a dockerfile
for RELEASE in $LIBVIPS_RELEASES;
do
  echo 'Release: \n' + $RELEASE
  # create a directory for the version
  # for each asset in the $RELEASE
  DESCRIPTION="$($RELEASE|jq '.body' $1))"
  VERSION_TAG="$($RELEASE|jq '.tag_name' $1))"
  mkdir -p ../tags/$($VERSION_TAG)/
  for asset in $(jq -s '.assets[]' $RELEASE);
  do
    # create a new dockerfile from $TMPL_PATH
    file=$($asset|jq '.name' $1).dockerfile
    destination="../src/$($VERSION_TAG)/$($file)"
    cp $TMPL_PATH $destination
    cat $destination | envsubst < [$DESCRIPTION $VERSION_TAG]
  done

done
exit 0
```

This is using `jq` to parse JSON on the shell and `envsubst` to do the template compiling. It's currently *not* working but I will edit this later once I get it to create the files. The problem with this approach is I'm just *assuming* everything will work just fine for each release. I also don't have the historical knowledge of each release except what is in git. My gut tells me there is probably a better tool for this, but if not we can write a small shell script to get there.

Once we have that we can add pushing to dockerhub as part of the normal build pipeline.

**Comment by toadkicker** [comment]
Created a repo: https://github.com/toadkicker/libvips-docker

**Comment by aried3r** [comment]
There is a GitHub action that can do this.
https://github.com/marketplace/actions/publish-docker

For #987 we could start with this action:
https://github.com/actions/upload-release-asset
Which could build linux, windows, macOS and put it into GitHub releases.

For deployment to Launchpad (not docker), this could be used on Travis:
https://docs.travis-ci.com/user/deployment/launchpad/
Since Travis now also runs on 18.04 if necessary.

**Comment by Nakilon** [comment]
While Dockerhub is a good thing and Github Actions is interesting I won't recommend to automate dockerhub pushing. Publishing it as an official docker image is the last step that should be 1) after the Github release 2) done manually 3) done carefully. Maybe only if it's automated with some tag like `-dev`.

**Comment by Nakilon** [comment]
Spent these ours trying out the Github Actions. Cool thing in general but... now when I've almost reached the step of publishing to Dockerhub I've discovered that Dockerhub's access token is not configurable, at all. You can't limit permissions or even images. So ideally you'll need to create some fake Dockerhub user to make it a "Collaborator" of a "Dockerhub repository".
Another approach would be to integrate it from another side -- https://docs.docker.com/docker-hub/builds/ -- I didn't explore it yet and probably won't. But what I see right now it wants you to authorize them at Github for access both for Public and Private repositories. So ideally you'll need to create a fake Github user.
So both ways are clumsy in this aspect.

**Comment by Nakilon** [comment]
Since then, in March, Docker made some relevant Github Action here: https://github.com/marketplace/actions/build-and-push-docker-images

But it's now also possible to host images in Github: https://help.github.com/en/actions/language-and-framework-guides/publishing-docker-images

And here is something about using it afterwards: https://help.github.com/en/packages/using-github-packages-with-your-projects-ecosystem/configuring-docker-for-use-with-github-packages#publishing-a-package

**Comment by lovell** [comment]
(A word of warning against using GitHub's Package Registry for open source software - it requires authentication even for public reads - see https://github.community/t5/GitHub-Actions/docker-pull-from-public-GitHub-Package-Registry-fail-with-quot/td-p/32782 for a stream of people coming to a similar conclusion.)
